[GTK] Remove key event reinjection

mcatanzaro opened the pull request:
#### fd1f0b783bb7b0b957f17c4e9594f11747e2be33
<pre>
[GTK] Remove key event reinjection
<a href="https://bugs.webkit.org/show_bug.cgi?id=261348">https://bugs.webkit.org/show_bug.cgi?id=261348</a>

Reviewed by Carlos Garcia Campos.

Event processing in GTK is synchronous, but in WebKit it is asynchronous
because we don&apos;t want to block the UI process waiting for the web
process to decide whether a DOM event has been handled (e.g. by using
Event.stopPropagation). Currently we use a complicated scheme to
synthesize and reinject new key and wheel events to continue event
processing when the event is not handled by the web content, which the
GTK developers do not approve of, and which is causing serious problems
where Epiphany has to choose between processing events in an infinite
loop vs. handling events too soon and blocking the web view from
receiving them. Neither option is great.

<a href="https://gitlab.gnome.org/GNOME/epiphany/-/issues/1915">https://gitlab.gnome.org/GNOME/epiphany/-/issues/1915</a>
<a href="https://gitlab.gnome.org/GNOME/epiphany/-/issues/2173">https://gitlab.gnome.org/GNOME/epiphany/-/issues/2173</a>

The solution is to just never reinject events, and instead always handle
them. We do not need event reinjection anymore if we activate
application accelerators manually. See the long comment embedded in this
commit for full details of why this is necessary.

This patch removes the key event reinjection. Wheel event reinjection is
not yet removed because this would break when WebKit is used within
scrolled windows.

 * Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
 (WebKit::PageClientImpl::doneWithKeyEvent):
 * Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
 (shouldForwardKeyEvent):
 (webkitWebViewBaseProcessAcceleratorsForKeyPressEvent):
 (webkitWebViewBaseKeyPressed):
 (handleScroll):
 (webkitWebViewBasePropagateKeyEvent):
 * Source/WebKit/UIProcess/API/gtk/WebKitWebViewBasePrivate.h:

Canonical link: <a href="https://commits.webkit.org/273922@main">https://commits.webkit.org/273922@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/64506ad4b075c264f5744b192e126f7b09c7b40d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37202 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16104 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39498 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39772 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33197 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18684 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13247 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37765 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13590 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/32738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11833 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41029 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33690 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/33820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37734 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12167 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9955 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35877 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13783 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8401 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12521 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/13045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -1133,18 +1097,58 @@ static void webkitWebViewBaseFocusLeave(WebKitWebViewBase* webViewBase, GtkEvent
+        // activate actions with no parameters, and I think it probably doesn't make sense for an
+        // action associated with an accelerator to have any parameters, but we should make sure.
+        // But I don't know how to do this.
+        gtk_widget_activate_action(GTK_WIDGET(webViewBase), actions.get()[i], nullptr);
```
This is the only reason this pull request is still a draft.

We will probably change the precondition of this function and remove the FIXME, because I think it's impossible to comply with the precondition and we think it should be fine: GTK will just print a warning if the action does require parameters, and I think it would be an application bug to register accelerators on an action that requires parameters.

mcatanzaro replied:
We found a way to pass along the correct parameter after all.

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -1133,18 +1128,60 @@ static void webkitWebViewBaseFocusLeave(WebKitWebViewBase* webViewBase, GtkEvent
+    // only after determining that the web process has not handled the event.) It's safe to do this
+    // even if the window already processed the events before WebKit, because in that case, no
+    // accelerator was activated (or it would have handled the event), so we'll fail to find any
+    // matching accelerator here. This clever strategy was proposed by Benjamin Otte.
```
What happens with other key events that might be handled by the application that are not accelerators?

mcatanzaro replied:
They will break, because the web view will handle the events unconditionally.

However, in practice I think it's fairly unlikely that this will affect real applications. Hopefully.

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -1133,18 +1128,60 @@ static void webkitWebViewBaseFocusLeave(WebKitWebViewBase* webViewBase, GtkEvent
+
+    GUniquePtr<char> accelerator(gtk_accelerator_name(gdk_key_event_get_keyval(event), gdk_event_get_modifier_state(event)));
+    GUniquePtr<char*> actions(gtk_application_get_actions_for_accel(GTK_APPLICATION(app), accelerator.get()));
+    for (int i = 0; actions.get()[i] != nullptr; i++) {
```
Don't compare to nullptr. i++ -> ++i

mcatanzaro replied:
OK.

That said, after 10 years of working together, only now to discover that you like `++i`... heartbreaking!

carlosgcampos reviewed (approved):
Ok, let's try this.

mcatanzaro committed:
[GTK] Remove key event reinjection
https://bugs.webkit.org/show_bug.cgi?id=261348

Reviewed by Carlos Garcia Campos.

Event processing in GTK is synchronous, but in WebKit it is asynchronous
because we don't want to block the UI process waiting for the web
process to decide whether a DOM event has been handled (e.g. by using
Event.stopPropagation). Currently we use a complicated scheme to
synthesize and reinject new key and wheel events to continue event
processing when the event is not handled by the web content, which the
GTK developers do not approve of, and which is causing serious problems
where Epiphany has to choose between processing events in an infinite
loop vs. handling events too soon and blocking the web view from
receiving them. Neither option is great.

https://gitlab.gnome.org/GNOME/epiphany/-/issues/1915
https://gitlab.gnome.org/GNOME/epiphany/-/issues/2173

The solution is to just never reinject events, and instead always handle
them. We do not need event reinjection anymore if we activate
application accelerators manually. See the long comment embedded in this
commit for full details of why this is necessary.

This patch removes the key event reinjection. Wheel event reinjection is
not yet removed because this would break when WebKit is used within
scrolled windows.

 * Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
 (WebKit::PageClientImpl::doneWithKeyEvent):
 * Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
 (shouldForwardKeyEvent):
 (webkitWebViewBaseProcessAcceleratorsForKeyPressEvent):
 (webkitWebViewBaseKeyPressed):
 (handleScroll):
 (webkitWebViewBasePropagateKeyEvent):
 * Source/WebKit/UIProcess/API/gtk/WebKitWebViewBasePrivate.h:

Canonical link: https://commits.webkit.org/273922@main
